### PR TITLE
Pass by reference to avoid out-of-bounds reads

### PR DESCRIPTION
**Pull request by naoyam**
This code results in out-of-bounds gmem accesses:

```
TEST(NVFuserTest, TMP) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  auto tv0 = makeSymbolicTensor(2);
  fusion.addInput(tv0);
  auto tv1 = sum(tv0, {1});
  fusion.addOutput(tv1);

  tv1->split(1, 128);

  tv1->axis(-1)->parallelize(ParallelType::TIDx);

  fusion.printMath();
  fusion.printKernel();

  int numel_x = 1;
  int numel_y = 129;

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  at::Tensor t0 = at::randn({numel_x, numel_y}, options);
  std::vector<IValue> inputs = {t0};

  FusionExecutor fe;
  fe.compileFusion(&fusion);
  auto outputs = fe.runFusion(inputs);

  auto t1 = t0.sum({1});

  testValidate(&fusion, outputs, inputs, {t1}, __LINE__, __FILE__);
}
```

```
__global__ void CUDAGeneratedKernel(Tensor<float, 2> T0, Tensor<float, 1> T1) {
  alignas(4) extern __shared__ char array[];
  void* shared_mem = array;
  #pragma unroll 1
  for(size_t ki23 = 0; ki23 < T0.size[0]; ++ki23) {
    T1[(ki23 * 1)] = 0;
  }
  #pragma unroll 1
  for(size_t ki19 = 0; ki19 < T0.size[0]; ++ki19) {
    #pragma unroll 1
    for(size_t ki20 = 0; ki20 < (ceilDiv(T0.size[1], 128)); ++ki20) {
      blockReduce<true, false, false>(
        T1[(ki19 * 1)],
        T0[(ki19 * T0.stride[0]) + (((ki20 * 128) + threadIdx.x) * T0.stride[1])],
        [](float &a, float b) { a = a + b; },
        threadIdx,
        blockDim,
        static_cast<float*>(shared_mem),
        (((ki20 * 128) + threadIdx.x) < T0.size[1]),
        float(0));
    }
  }
}
```

The problem is `blockReduce` has an argument, `T0[(ki19 * T0.stride[0]) + (((ki20 * 128) + threadIdx.x) * T0.stride[1])]`, which needs to be predicated. It is indeed predicated within the `blockReduce` function, but since it is passed by value, the read access should happen before it is used in `blockReduce`. The illegal access can be detected by compute-sanitizer with `PYTORCH_NO_CUDA_MEMORY_CACHING=1`. 

This PR avoids the problem by passing the input values by reference.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/runtime/welford.cu [hunk 1]
```diff
@@ -353,9 +353,9 @@ __device__ bool gridWelford(
     T& out_M2,
     T& out_avg,
     TN& out_N,
-    T inp_M2,
-    T inp_avg,
-    TN inp_N,
+    const T& inp_M2,
+    const T& inp_avg,
+    const TN& inp_N,
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/runtime/welford.cu:358]: @naoyam Thanks for helping fixing this. There seems to be another one at line40, in `blockWelford`.

**Review by shmsong — APPROVED** [review]
LGTM.

File: torch/csrc/jit/codegen/cuda/runtime/welford.cu (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/runtime/welford.cu:358]: Changed it to a const reference. Thanks.

**Comment by csarofeen** [comment]
Good catch.

**Pull request merged**